security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -279,7 +279,12 @@ func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecre
 			// Reset SDS push time for new SDS push.
 			con.sdsPushTime = time.Time{}
 			con.mutex.Unlock()
-			defer recycleConnection(conID, resourceName)
+			defer func() {
+				recycleConnection(conID, resourceName)
+				// Remove the secret from cache, otherwise refresh job will process this item(if envoy fails to reconnect)
+				// and cause some confusing logs like 'fails to notify because connection isn't found'.
+				s.st.DeleteSecret(conID, resourceName)
+			}()
```

Comment:
Looks like we have done this at line 385 when sds-agent tries to close a connection. Given that s.st.DeleteSecret() should always follow recycleConnection(), can we define a new method to wrap these operations? 
releaseResourcePerConn(conID, resourceName) {
   recycleConnection(conID, resourceName)
   s.st.DeleteSecret(conID, resourceName)
}
In the future, other cleanup operations should be added into releaseResourcePerConn(), this also helps reduces the size of StreamSecrets().

Reply:
Yes refactor is required